### PR TITLE
fixing unfolding again

### LINKED ISSUE #139: not unfolding somewhere?
```
def path (A : type) (x : A) (y : A) : type =
  pathd {_ => A} x y

def test (A : type) (φ : 𝔽) (a : A)
  (p : sub {path A a a} φ {_ => a})
  : (j : 𝕀) → A
  =
  p
```
raises
```
cooltt: internal error, uncaught exception:
        Head connective mismatch, expected pi but got path A a a
```
It works if I inline `path` or remove the `sub`.

### PR DESCRIPTION
**Pull request by jonsterling**
Third time's the charm lol

Resolves #139

### CODE REVIEW

File: src/lib/Semantics.ml [hunk 1]
```diff
@@ -184,7 +181,7 @@ let rec cof_con_to_cof : (D.con, D.con) Cof.cof_f -> D.cof CM.m =
 and con_to_cof =
   let open CM in
   fun con ->
-    whnf_inspect_con con |>>
+    whnf_inspect_con ~style:`UnfoldAll con |>>
```
Comment (jonsterling) [inline src/lib/Semantics.ml:184]: something I didn't catch before!

File: src/lib/Elaborator.ml [hunk 2]
```diff
@@ -160,6 +160,9 @@ and bchk_tm : CS.con -> T.BChk.tac =
   match con.node with
   | CS.Hole name ->
     R.Hole.unleash_hole name `Rigid
+  | CS.Unfold (idents, c) ->
```
Comment (jonsterling) [inline src/lib/Elaborator.ml:163]: Exercise to test the reader's Elaboration Geometry

**Comment by ecavallo** [comment]
I'm still getting the same error for the example in #139... is this just me, am I failing to run the right cooltt somehow?

**Comment by jonsterling** [comment]
@ecavallo I see... I get it too. This is a separate issue that is orthogonal to unfolding, I believe. I will look into it.

**Comment by jonsterling** [comment]
Or maybe it is not fully orthogonal lol. But I will look into it nonetheless.

**Comment by jonsterling** [comment]
OK @ecavallo I tracked down the issue, thank you for checking before! I've added your example to the test to prove that I really fixed it this time LOL 😆

File: src/lib/Semantics.ml [hunk 3]
```diff
@@ -857,11 +854,20 @@ and plug_into ~style sp con =
   | `Done -> ret @@ `Reduce res
   | `Reduce res -> ret @@ `Reduce res
 
-and whnf_hd ?(style = default_whnf_style) hd =
+and should_unfold_symbol style sym =
+  match style with
+  | `UnfoldNone -> false
+  | `UnfoldAll -> true
+  | `Veil veil ->
+    match Veil.policy sym veil with
```
Comment (favonia) [inline src/lib/Semantics.ml:862]: I wonder if it makes sense to update the veil so that it only unfold <= N times or maybe only under <= M binders LOL.

Comment (jonsterling) [reply src/lib/Semantics.ml:862]: LOL. I am open to many different Cool ideas for veil policies haha. Though I think that one may be a little fragile.

Anyway, we can make a Cool Mini-Language for unfolding if we are board ;-)

**Comment by ecavallo** [comment]
Hm, this doesn't fix the original problem I was having though. I'll try to strip it down a bit, but for reference:
```
[...contents of test/groupoid-laws...]

def test (A : type)
  (p : (i : 𝕀) → A)
  (q : (i : 𝕀) → sub A {i=0} {p 1})
  (r : (i : 𝕀) → sub A {i=0} {q 1})
  (s : (i : 𝕀) → sub A {i=0} {r 1})
  : (j : 𝕀) → path A {p 0} {r 1}
  =
  j => assoc A p #f q #f r j
```
gives
```
cooltt: internal error, uncaught exception:
        Head connective mismatch, expected pi but got path {path A {p 0} {r 1}} {
        _x₁ =>
        trans A {_x₂ => trans A {_x₃ => p _x₃} #f {_x₃ => q _x₃} _x₂} #f {_x₂ =>
                                                                          r _x₂} _x₁} {
        _x₁ =>
        trans A {_x₂ => p _x₂} #f {_x₂ =>
                                   trans A {_x₃ => q _x₃} #f {_x₃ => r _x₃} _x₂} _x₁}
```
It checks if I eta-contract.

**Comment by jonsterling** [comment]
Thank you for the more sophisticated example; it is necessary to exercise this code more strenuously it would seem LOL

**Comment by ecavallo** [comment]
Same for this simpler one
```
def path (A : type) (x : A) (y : A) : type =
  pathd {_ => A} x y

def refl2 (A : type) (p : (i : 𝕀) → A) : sub {path {path A {p 0} {p 1}} p p} #t {_ => p}
  =
  _ => p

def test (A : type) (p : (i : 𝕀) → A) : (j : 𝕀) → path A {p 0} {p 1}
  =
  j => refl2 A p j
```

**Comment by jonsterling** [comment]
@ecavallo OK, how is it now?

File: src/lib/Semantics.ml (hunk 3 above)
Comment (favonia) [reply src/lib/Semantics.ml:862]: Let's leave my Immortal Thought in another thread...

**Comment by ecavallo** [comment]
Looks good!

**Comment by jonsterling** [comment]
@ecavallo Thanks! I don't know what we would do without your trademarked Counterexample Generator lol :)

**Pull request merged**